### PR TITLE
Optimize setNode

### PR DESCRIPTION
**Pull request by dapplion**
**Motivation**

Keeping a record of the parent nodes in an array instead of binded functions is faster.

**Description**

Refactor setNode to:
- Don't use iterators. Calculate entire bitstring at once
- Keep a list of parentNodes, then iterate it backwards to replace nodes.

Overall the result is about x2 faster

```
  set
    ✓ set at depth 8                             1607717 ops/s    622.0000 ns/op   x0.556    1057567 runs   1.05 s
    ✓ set at depth 16                            1254705 ops/s    797.0000 ns/op   x0.449     967763 runs   1.04 s
    ✓ set at depth 32                           882612.5 ops/s    1.133000 us/op   x0.425     731535 runs   1.03 s
```

### CODE REVIEW

File: src/tree.ts [hunk 1]
```diff
@@ -71,38 +71,51 @@ export class Tree {
...
-      if (i) {
-        if (node.isLeaf()) {
-          if (!expand) throw new Error(ERR_INVALID_TREE);
-          else {
-            const child = zeroNode(iterator.remainingBitLength() - 1);
-            node = new BranchNode(child, child);
-          }
+
+    // Pre-compute entire bitstring instead of using an iterator (25% faster)
+    const bitstring = index.toString(2);
```
Comment (wemeetagain) [inline src/tree.ts:78]: Can you check the index < 1 case and add unit test? Previously handled by gindexIterator

**Review by wemeetagain — CHANGES_REQUESTED** [review]
:heart:
approach LGTM!

File: src/tree.ts [hunk 2]
```diff
@@ -71,38 +71,54 @@ export class Tree {
...
+    const parentNodes: Node[] = [this.rootNode];
+
+    // Ignore the first bit, left right directions are at bits [1,..]
+    // Ignore the last bit, no need to push the target node to the parentNodes array
+    for (let i = 1; i < bitstring.length - 1; i++) {
+      if (node.isLeaf()) {
+        if (!expand) {
+          throw new Error(ERR_INVALID_TREE);
+        } else {
+          node = zeroNode(bitstring.length - i);
```
Comment (dapplion) [inline src/tree.ts:94]: @wemeetagain Can you review this line? How could I test this code path properly?

Comment (wemeetagain) [reply src/tree.ts:94]: This is for when setting a node at a depth thats further than the existing depth.
The tree would be expanded and filled w zero nodes.

This is what the current behavior looked like:
```
// Lets say we want to insert node n into this tree at index 8
            x
       /         \
      x           x

---
// step one is to expand into zero nodes

            x
       /         \
      z2           x
   /      \
  z1       z1
/    \    /   \
z0   z0  z0   z0

---
// Then replace node at index 8

            x
       /         \
      x           x
   /      \
  x       z1
/    \   /   \
n   z0  z0   z0
```

So to test, you can create a short tree and set a node deeper than the bottom.

Comment (wemeetagain) [reply src/tree.ts:99]: I'm having trouble creating a test even on master. The behavior seems to be the same between branches tho. I've tested this branch against lodestar, syncing various networks. Will figure out how to test this in a future PR.

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**